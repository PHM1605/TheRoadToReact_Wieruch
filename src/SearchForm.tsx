import React from 'react';
import { InputWithLabel } from "./InputWithLabel";

type SearchFormProps = {
  searchTerm:string; 
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

const SearchForm = (
    {searchTerm, onSearchInput, onSearchSubmit}: SearchFormProps) => (
    //console.log("RENDERING SEARCH FORM") ||
    <form onSubmit={onSearchSubmit} className="searchForm">
      <InputWithLabel id="search" value={searchTerm} isFocused onInputChange={onSearchInput}>
        Search:
      </InputWithLabel>
      <button type="submit" disabled={!searchTerm} className="button buttonLarge">Submit</button>
    </form>
);
export default SearchForm;