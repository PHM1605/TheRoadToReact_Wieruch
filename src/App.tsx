import React from 'react';
import './App.css';
import axios from 'axios';
import SearchForm from './SearchForm';
import List from './List';

const API_BASE ='https://hn.algolia.com/api/v1';
const API_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';

type Story = {
  objectID: string, url: string, title: string, author: string, num_comments: number, points: number
};
type Stories = Array<Story>;

const getSumComments = (stories) => {
  // console.log('C');
  return stories.data.reduce((result, value) => result + value.num_comments, 0);
};

type StoriesState = {
  data: Stories;
  page: number;
  isLoading: boolean;
  isError: boolean;
};



interface StoriesFetchInitAction {
  type: 'STORIES_FETCH_INIT';
}

interface StoriesFetchSuccessAction {
  type: 'STORIES_FETCH_SUCCESS';
  payload: {
    list: Stories;
    page: number;
  }
}

interface StoriesFetchFailureAction {
  type: 'STORIES_FETCH_FAILURE';
}

interface StoriesRemoveAction {
  type: 'REMOVE_STORY';
  payload: Story;
}

type StoriesAction = 
  | StoriesFetchInitAction 
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | StoriesRemoveAction;

const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch(action.type) {
    case 'STORIES_FETCH_INIT':
      return {...state, isLoading: true, isError: false};
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state, 
        isLoading: false, 
        isError: false, 
        data: action.payload.page ===0 ? action.payload.list : state.data.concat(action.payload.list), 
        page: action.payload.page};
    case 'STORIES_FETCH_FAILURE':
      return {...state, isLoading: false, isError: true};
    case 'REMOVE_STORY':
      return {...state, 
        data: state.data.filter((story) => action.payload.objectID !== story.objectID)};
    default:
      throw new Error();
  }
};

const useSemiPersistentState = (key:string, initialState:string):
[string, (newValue: string)=>void] => {
  const isMounted = React.useRef(false);
  const [value, setValue] = React.useState(localStorage.getItem(key) || initialState);
  React.useEffect(()=>{
    if (!isMounted.current) {
      isMounted.current = true;
    } else {
      localStorage.setItem(key, value);
    }
  }, [value, key]);
  return [value, setValue];
};

const extractSearchTerm = (url) => url
  .substring(url.lastIndexOf('?')+1, url.lastIndexOf('&'))
  .replace(PARAM_SEARCH, '');
const getLastSearches = (urls) => urls
  .reduce((result, url, index)=>{
    const searchTerm = extractSearchTerm(url);
    if(index===0) {
      return result.concat(searchTerm);
    }
    const previousSearchTerm = result[result.length - 1];
    if (searchTerm === previousSearchTerm) {
      return result;
    } else {
      return result.concat(searchTerm);
    }
  }, [])
  .slice(-6).slice(0, -1);

const LastSearches = ({lastSearches, onLastSearch}) => (
  <>
    {lastSearches.map((searchTerm, index)=> (
      <button key={searchTerm+index} type="button" onClick={()=>onLastSearch(searchTerm)}>
        {searchTerm}
      </button>
    )
    )}
  </>
);

const App = () => {
  const getUrl = (searchTerm, page) => 
    `${API_BASE}${API_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`;
  const [searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');
  const [urls, setUrls] = React.useState([getUrl(searchTerm, 0)]);
  const [stories, dispatchStories] = React.useReducer(storiesReducer, {data: [], page:0, isLoading: false, isError: false});
  
  const handleSearchInput = React.useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  }, [setSearchTerm]);

  const handleSearch = (searchTerm, page) => {
    const url = getUrl(searchTerm, page);
    setUrls(urls.concat(url));
  };
  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    handleSearch(searchTerm, 0);
    event.preventDefault();
  };
  const handleLastSearch = (searchTerm) => {
    setSearchTerm(searchTerm);
    handleSearch(searchTerm, 0);
  };

  const handleFetchStories = React.useCallback(async ()=>{
    dispatchStories({type: 'STORIES_FETCH_INIT'});
    try {
      const lastUrl = urls[urls.length - 1];
      const result = await axios.get(lastUrl);
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS', 
        payload: {
          list: result.data.hits,
          page: result.data.page,
        }
      });
    } catch {
      dispatchStories({type: 'STORIES_FETCH_FAILURE'});
    }
  }, [urls]);

  React.useEffect(()=>{
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = React.useCallback((item: Story) => {
    dispatchStories({type: 'REMOVE_STORY', payload: item});
  }, []);
  
  const handleMore = () => {
    const lastUrl = urls[urls.length-1];
    const searchTerm = extractSearchTerm(lastUrl);
    handleSearch(searchTerm, stories.page + 1);
  }

  //console.log('B:App');
  const sumComments = React.useMemo(() => getSumComments(stories), [stories]);
  const lastSearches = getLastSearches(urls);
  return (
    <div className='container'>
      <h1 className='headline-primary'>My Hacker Stories with {sumComments} comments.</h1>

      <SearchForm searchTerm={searchTerm} onSearchInput={handleSearchInput} onSearchSubmit={handleSearchSubmit}/>
      <LastSearches lastSearches={lastSearches} onLastSearch={handleLastSearch} />

      {stories.isError && <p>Something went wrong...</p>}
      <List list={stories.data} onRemoveItem={handleRemoveStory}/>
      {stories.isLoading ? (
        <p>Loading ...</p>
      ) : (
        <button type="button" onClick={handleMore}>
          More 
        </button>
      )}
    </div>
  );
};
export default App;
export {Story, Stories, storiesReducer};