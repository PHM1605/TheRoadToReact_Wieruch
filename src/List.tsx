import React from 'react';
import check_img from './check.png';
import {Story, Stories} from './App.tsx';
import {sortBy} from 'lodash';

type ItemProps = {item: Story; onRemoveItem: (item:Story)=>void;};
type ListProps = {list: Stories; onRemoveItem: (item:Story)=>void;};

const SORTS = {
  NONE: (list) => list,
  TITLE: (list) => sortBy(list, 'title'),
  AUTHOR: (list) => sortBy(list, 'author'),
  COMMENT: (list) => sortBy(list, 'num_comments').reverse(),
  POINT: (list) => sortBy(list, 'points').reverse()
};

const List = ({list, onRemoveItem}: ListProps) => {
  const [sort, setSort] = React.useState({
    sortKey: 'NONE',
    isReverse: false
  });

  const handleSort = (sortKey) => {
    // if pressing the same button AND currently not reverse
    const isReverse = sort.sortKey === sortKey && !sort.isReverse;
    setSort({sortKey: sortKey, isReverse: isReverse});
  };
  const sortFunction = SORTS[sort.sortKey];
  const sortedList = sort.isReverse ? sortFunction(list).reverse() : sortFunction(list);
  //console.log('B:List') ||
  return (
    <ul>
      <li style={{display: 'flex'}}>
        <span style={{width: '40%'}}>
          <button type="button" onClick={()=>handleSort('TITLE')}>Title</button>
        </span>
        <span style={{width: '30%'}}>
          <button type="button" onClick={()=>handleSort('AUTHOR')}>Author</button>
        </span>
        <span style={{width: '10%'}}>
          <button type="button" onClick={()=>handleSort('COMMENT')}>Comments</button>
        </span>
        <span style={{width: '10%'}}>
          <button type="button" onClick={()=>handleSort('POINT')}>Points</button>
        </span>
        <span style={{width: '10%'}}>
          Action
        </span>
      </li>
      {sortedList.map((item)=>{
        return (
        <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
        );
      })
      }
    </ul>
  );
};
  
const Item = ({item, onRemoveItem}: ItemProps) => (
    <li className="item">
        <span style={{width:"40%"}}><a href={item.url}>{item.title}</a></span>
        <span style={{width:"30%"}}>{item.author}</span>
        <span style={{width:"10%"}}>{item.num_comments}</span>
        <span style={{width:"10%"}}>{item.points}</span>
        <span style={{width:"10%"}}>
        <button type="button" onClick={()=>onRemoveItem(item)} className="button buttonSmall">
            <img src={check_img} alt="check" height="18px" width="18px" />
        </button>
        </span>
    </li>
);

export default List;
export {Item};