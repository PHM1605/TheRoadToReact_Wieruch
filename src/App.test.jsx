import { describe, it, expect, vi} from 'vitest';
import {render, screen, fireEvent, waitFor} from '@testing-library/react';
import App, {storiesReducer} from './App.tsx';
import {Item} from './List.tsx';
import SearchForm from './SearchForm.tsx';

import axios from 'axios';

vi.mock('axios');

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  };
  
const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};
  
const stories = [storyOne, storyTwo];

describe('storiesReducer', () => {
it('removes a story from all stories', () => {
        const action = { type: 'REMOVE_STORY', payload: storyOne };
        const state = { data: stories, isLoading: false, isError: false };
        const newState = storiesReducer(state, action);
        const expectedState = {
            data: [storyTwo],
            isLoading: false,
            isError: false,
        };
        expect(newState).toStrictEqual(expectedState);
    });
});

describe('Item', ()=>{
    it('renders all properties', ()=>{
        render(<Item item={storyOne}/>);
        // screen.debug();
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getByText('React')).toHaveAttribute('href', 'https://reactjs.org/');
    });
    it('renders a clickable dismiss button', ()=>{
        render(<Item item={storyOne}/>);
        expect(screen.getByRole('button')).toBeInTheDocument();
    });
    it('clicking the dismiss button calls the callback handler', ()=>{
        const handleRemoveItem = vi.fn();
        render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);
        fireEvent.click(screen.getByRole('button'));
        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    });
});

describe('SearchForm', ()=>{
    const searchFormProps = {
        searchTerm: 'React', 
        onSearchInput: vi.fn(), 
        onSearchSubmit:vi.fn()
    };
    it('renders the input field with its value', ()=>{
        render(<SearchForm {...searchFormProps} />);
        expect(screen.getByDisplayValue('React')).toBeInTheDocument();
    });
    it('renders the correct label', ()=>{
        render(<SearchForm {...searchFormProps} />);
        expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
    });
    it('calls onSearchInput on input field change', ()=>{
        render(<SearchForm {...searchFormProps} />);
        fireEvent.change(screen.getByDisplayValue('React'), {target: {value: 'Redux'}});
        expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
    });
    it('renders snapshot', ()=>{
        const {container} = render(<SearchForm {...searchFormProps}/>);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('App', ()=>{
    it('succeeds fetching data', async ()=>{
        const promise = Promise.resolve({
            data:{
                hits: stories
            }
        });
        axios.get.mockImplementationOnce(()=>promise);
        render(<App />);
        //screen.debug();
        expect(screen.queryByText(/Loading/)).toBeInTheDocument();
        await waitFor(async ()=>await promise);
        //screen.debug();
        expect(screen.queryByText(/Loading/)).toBeNull();
        expect(screen.getByText('React')).toBeInTheDocument();
        expect(screen.getByText('Redux')).toBeInTheDocument();
        expect(screen.getAllByRole('img').length).toBe(2);
    });

    it('fails fetching data', async ()=>{
        const promise = Promise.reject();
        axios.get.mockImplementationOnce(()=>promise);
        render(<App/>);
        expect(screen.getByText(/Loading/)).toBeInTheDocument();
        try {
            await waitFor(async ()=> await promise);
        } catch(error) {
            expect(screen.queryByText(/Loading/)).toBeNull();
            expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
        }
    });

    it('removes a story', async ()=>{
        const promise = Promise.resolve({data: {hits: stories}});
        axios.get.mockImplementationOnce(()=>promise);
        render(<App/>);
        await waitFor(async () => await promise);
        expect(screen.getAllByRole('img').length).toBe(2);
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        // fire event delete and check again
        fireEvent.click(screen.getAllByRole('img')[0]);
        expect(screen.getAllByRole('img').length).toBe(1);
        expect(screen.queryByText('Jordan Walke')).toBeNull();
    });

    it('searches for specific stories', async()=>{
        const reactPromise = Promise.resolve({
            data: {
                hits: stories,
            }
        });
        const anotherStory = {
            title: 'JavaScript',
            url: 'https://en.wikipedia.org/wiki/JavaScript',
            author: 'Brendan Eich',
            num_comments: 15,
            points: 10,
            objectID: 3
        };
        const javascriptPromise = Promise.resolve({
            data: {
                hits: [anotherStory],
            }
        });

        axios.get.mockImplementation((url)=>{
            if(url.includes('React')) {
                return reactPromise;
            }
            if(url.includes('JavaScript')) {
                return javascriptPromise;
            }
            throw Error();
        });

        render(<App/>);
        // Checking the first Promise for initial render
        await waitFor(async ()=> await reactPromise);
        expect(screen.queryByDisplayValue('React')).toBeInTheDocument(); // check input value to be 'React'
        expect(screen.queryByDisplayValue('JavaScript')).toBeNull(); // initial list does not contain 'Javascript' yet
        expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.queryByText('Dan Abramov, Andrew Clark')).toBeInTheDocument();
        expect(screen.queryByText('Brendan Eich')).toBeNull();

        // Checking user interaction via typing in the Input field
        fireEvent.change(screen.queryByDisplayValue('React'), { target: {value: 'JavaScript'}});
        expect(screen.queryByDisplayValue('React')).toBeNull();
        expect(screen.queryByDisplayValue('JavaScript')).toBeInTheDocument();

        // Checking user interaction via submit
        fireEvent.submit(screen.queryByText('Submit'));
        await waitFor(async ()=> await javascriptPromise);
        //screen.debug();
        expect(screen.queryByText('Jordan Walke')).toBeNull();
        expect(screen.queryByText('Dan Abramov, Andrew Clark')).toBeNull();
        expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();
    });
});