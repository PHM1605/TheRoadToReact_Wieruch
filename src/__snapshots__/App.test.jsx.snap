// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="searchForm"
>
  <label
    class="label"
    for="search"
  >
    Search:
  </label>
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button buttonLarge"
    type="submit"
  >
    Submit
  </button>
</form>
`;
